File: ghr_custom/ghr_custom/report/party_ledger_statement/party_ledger_statement.js
frappe.query_reports["Party Ledger Statement"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_start(),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_end(),
		},
		{
			fieldname: "party_type",
			label: __("Party Type"),
			fieldtype: "Link",
			options: "Party Type",
			on_change: function (report) {
				const party_type = report.get_filter_value("party_type");
				const party_filter = report.get_filter("party");

				report.set_filter_value("party", "");

				if (party_type) {
					party_filter.df.options = party_type;
				} else {
					party_filter.df.options = "";
				}

				party_filter.refresh();
			},
		},
		{
			fieldname: "party",
			label: __("Party"),
			fieldtype: "Link",
			options: "",
			get_query: function () {
				const party_type = frappe.query_report.get_filter_value("party_type");
				const company = frappe.query_report.get_filter_value("company");

				if (!party_type || !company) {
					return {};
				}

				return {
					query: "ghr_custom.ghr_custom.report.party_ledger_statement.party_ledger_statement.get_party_query",
					filters: {
						party_type: party_type,
						company: company,
					},
				};
			},
		},
		{
			fieldname: "account",
			label: __("Account"),
			fieldtype: "Link",
			options: "Account",
			get_query: function () {
				const company = frappe.query_report.get_filter_value("company");
				let filters = { is_group: 0 };

				if (company) {
					filters.company = company;
				}

				return { filters: filters };
			},
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
			get_query: function () {
				const company = frappe.query_report.get_filter_value("company");
				let filters = {};

				if (company) {
					filters.company = company;
				}

				return { filters: filters };
			},
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "include_opening",
			label: __("Include Opening"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "show_narration",
			label: __("Show Remarks"),
			fieldtype: "Check",
			default: 1,
			on_change: function (report) {
				report.refresh();
			},
		},
	],

	get_datatable_options(options) {
		return Object.assign(options, {
			serialNoColumn: false,
			checkboxColumn: false,
			inlineFilters: false,
			dynamicRowHeight: true,
			cellHeight: 42,
		});
	},

	onload: function (report) {
		inject_party_ledger_styles();

		const party_type = report.get_filter_value("party_type");
		const party_filter = report.get_filter("party");

		if (party_type) {
			party_filter.df.options = party_type;
			party_filter.refresh();
		}

		report.page.add_inner_button(__("Validate Filters"), function () {
			validate_party_or_account(report);
		});

		report.page.add_inner_button(__("Print Ledger"), function () {
			print_party_ledger(report);
		});
	},

	before_refresh: function () {
		validate_party_or_account(frappe.query_report, false);
	},

	formatter: function (value, row, column, data, default_formatter) {
		let formatted_value = default_formatter(value, row, column, data);

		if (!data) return formatted_value;

		const row_kind = data.row_kind || "normal";

		if (column.fieldname === "particulars") {
			return data.particulars || "";
		}

		if (column.fieldname === "remarks") {
			if (!data.remarks) return "";
			return `<div class="pls-remarks">${frappe.utils.escape_html(data.remarks)}</div>`;
		}

		if (column.fieldname === "voucher_subtype" && value) {
			return `<div class="pls-subtype">${frappe.utils.escape_html(value)}</div>`;
		}

		if (column.fieldname === "balance") {
			let cls = "pls-balance";
			if (row_kind === "opening") cls += " pls-balance-opening";
			if (row_kind === "total") cls += " pls-balance-total";
			if (row_kind === "closing") cls += " pls-balance-closing";
			return `<div class="${cls}">${formatted_value}</div>`;
		}

		if (["debit", "credit"].includes(column.fieldname) && (row_kind === "total" || row_kind === "closing")) {
			return `<div class="pls-amount-strong">${formatted_value}</div>`;
		}

		return formatted_value;
	},
};

function validate_party_or_account(report, show_message = true) {
	const party = report.get_filter_value("party");
	const account = report.get_filter_value("account");

	if (!party && !account) {
		if (show_message) {
			frappe.msgprint(__("Please select either one Party or one Account."));
		}
		return false;
	}

	if (party && account) {
		if (show_message) {
			frappe.msgprint(__("Please select only one filter: either Party or Account, not both."));
		}
		return false;
	}

	return true;
}

function print_party_ledger(report) {
	const filters = report.get_values();

	if (!validate_party_or_account(report, true)) {
		return;
	}

	frappe.call({
		method: "ghr_custom.ghr_custom.report.party_ledger_statement.party_ledger_statement.get_print_data",
		args: {
			filters: filters,
		},
		callback: function (r) {
			if (!r.message) return;

			const printData = r.message;
			const html = build_party_ledger_print_html(printData);

			const printWindow = window.open("", "_blank");
			printWindow.document.open();
			printWindow.document.write(html);
			printWindow.document.close();

			setTimeout(() => {
				printWindow.focus();
				printWindow.print();
			}, 500);
		},
	});
}

function build_party_ledger_print_html(data) {
	const rowsHtml = build_print_rows_html(data.rows || []);

	return `
<!DOCTYPE html>
<html>
<head>
	<meta charset="utf-8">
	<title>Party Ledger Statement</title>
	<style>
		@page {
			size: A4 portrait;
			margin: 14mm 12mm 14mm 12mm;
		}

		body {
			font-family: Arial, Helvetica, sans-serif;
			color: #1f2937;
			margin: 0;
			padding: 0;
			background: #ffffff;
			font-size: 12px;
		}

		.print-wrap {
			width: 100%;
		}

		.header-card {
			border: 1px solid #d9e3f0;
			background: linear-gradient(135deg, #f8fbff 0%, #eef5ff 100%);
			border-radius: 14px;
			padding: 16px 18px;
			margin-bottom: 14px;
		}

		.company-name {
			font-size: 22px;
			font-weight: 700;
			color: #102a43;
			margin-bottom: 4px;
		}

		.report-title {
			font-size: 16px;
			font-weight: 700;
			color: #1f3b5b;
			margin-bottom: 10px;
		}

		.meta-grid {
			display: grid;
			grid-template-columns: 1fr 1fr;
			gap: 6px 18px;
			font-size: 12px;
			color: #44546a;
		}

		.meta-grid div strong {
			color: #243b53;
		}

		table {
			width: 100%;
			border-collapse: collapse;
			table-layout: fixed;
		}

		thead th {
			background: #eef2f7;
			color: #243b53;
			font-weight: 700;
			font-size: 12px;
			border: 1px solid #d8dee9;
			padding: 8px 8px;
			text-align: left;
		}

		tbody td {
			border: 1px solid #e5e7eb;
			padding: 7px 8px;
			vertical-align: top;
			font-size: 12px;
			word-wrap: break-word;
		}

		.text-right {
			text-align: right;
		}

		.part-main {
			font-weight: 600;
			color: #1f2937;
			line-height: 1.45;
		}

		.row-narration td {
			padding-top: 4px;
			padding-bottom: 7px;
			background: #ffffff;
		}

		.narration-span-cell {
			font-size: 11px;
			color: #6b7280;
			line-height: 1.45;
			font-style: italic;
			white-space: pre-wrap;
		}

		.row-opening td {
			background: #f8fbff;
		}

		.row-opening .part-main,
		.row-opening .balance-cell {
			font-weight: 700;
			color: #163b65;
		}

		.row-total td {
			background: #fff9f0;
			font-weight: 700;
		}

		.row-total .part-main,
		.row-total .balance-cell {
			color: #8a5200;
		}

		.row-closing td {
			background: #f3fbf6;
			font-weight: 700;
		}

		.row-closing .part-main,
		.row-closing .balance-cell {
			color: #0f5132;
		}

		.footer-note {
			margin-top: 10px;
			font-size: 10px;
			color: #6b7280;
			text-align: right;
		}
	</style>
</head>
<body>
	<div class="print-wrap">
		<div class="header-card">
			<div class="company-name">${escape_html_safe(data.company || "")}</div>
			<div class="report-title">Party Ledger Statement</div>
			<div class="meta-grid">
				<div><strong>Ledger:</strong> ${escape_html_safe(data.ledger_name || "")}</div>
				<div><strong>Type:</strong> ${escape_html_safe(data.party_type || "Account Ledger")}</div>
				<div><strong>From Date:</strong> ${escape_html_safe(data.from_date || "")}</div>
				<div><strong>To Date:</strong> ${escape_html_safe(data.to_date || "")}</div>
			</div>
		</div>

		<table>
			<thead>
				<tr>
					<th>Date</th>
					<th>Particulars</th>
					<th>Voucher Subtype</th>
					<th>Voucher No</th>
					<th class="text-right">Debit</th>
					<th class="text-right">Credit</th>
					<th class="text-right">Balance</th>
				</tr>
			</thead>
			<tbody>
				${rowsHtml}
			</tbody>
		</table>

		<div class="footer-note">Generated from ERPNext Party Ledger Statement</div>
	</div>
</body>
</html>
	`;
}

function build_print_rows_html(rows) {
	return rows
		.map((row) => {
			if ((row.row_kind || "normal") === "narration") {
				return `
					<tr class="row-narration">
						<td></td>
						<td colspan="3" class="narration-span-cell">${escape_html_safe(row.narration_only || "")}</td>
						<td></td>
						<td></td>
						<td></td>
					</tr>
				`;
			}

			const particulars = build_print_particulars(row);
			const rowClass = `row-${row.row_kind || "normal"}`;

			return `
				<tr class="${rowClass}">
					<td>${escape_html_safe(row.posting_date || "")}</td>
					<td>${particulars}</td>
					<td>${escape_html_safe(row.voucher_subtype || "")}</td>
					<td>${escape_html_safe(row.voucher_no || "")}</td>
					<td class="text-right">${format_print_amount(row.debit)}</td>
					<td class="text-right">${format_print_amount(row.credit)}</td>
					<td class="text-right balance-cell">${escape_html_safe(row.balance || "")}</td>
				</tr>
			`;
		})
		.join("");
}

function build_print_particulars(row) {
	const mainText = row.particulars_plain || "";
	return `<div class="part-main">${escape_html_safe(mainText)}</div>`;
}

function format_print_amount(value) {
	const num = flt(value);
	if (!num) return "";
	return format_currency(num, frappe.defaults.get_default("currency"));
}

function escape_html_safe(value) {
	if (value === null || value === undefined) return "";
	return String(value)
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&#39;");
}

function inject_party_ledger_styles() {
	if (document.getElementById("party-ledger-statement-styles")) return;

	const style = document.createElement("style");
	style.id = "party-ledger-statement-styles";
	style.innerHTML = `
		.pls-main {
			font-weight: 500;
			color: #1f2937;
			line-height: 1.45;
		}

		.pls-remarks {
			color: #6b7280;
			font-size: 11px;
			line-height: 1.5;
			font-style: italic;
			white-space: normal !important;
			word-break: break-word;
			overflow-wrap: anywhere;
			display: block;
			min-height: 18px;
		}

		.pls-special {
			font-weight: 700;
		}

		.pls-opening {
			color: #163b65;
		}

		.pls-total {
			color: #8a5200;
		}

		.pls-closing {
			color: #0f5132;
		}

		.pls-subtype {
			font-weight: 500;
			color: #4b5563;
		}

		.pls-balance {
			font-weight: 600;
			text-align: right;
		}

		.pls-balance-opening {
			color: #163b65;
		}

		.pls-balance-total {
			color: #8a5200;
		}

		.pls-balance-closing {
			color: #0f5132;
		}

		.pls-amount-strong {
			font-weight: 700;
		}

		.query-report .dt-row-header,
		.query-report .dt-cell--col-0 {
			display: none !important;
		}

		.query-report .dt-scrollable {
			border-radius: 16px;
			overflow: hidden;
			border: 1px solid #e5e7eb;
		}

		.query-report .dt-cell__content {
			padding-top: 10px;
			padding-bottom: 10px;
			white-space: normal !important;
			line-height: 1.45;
			overflow: visible !important;
			text-overflow: unset !important;
		}

		.query-report .datatable .dt-cell {
			vertical-align: top;
			overflow: visible !important;
		}

		.query-report .datatable .dt-cell__content--col-remarks,
		.query-report .datatable .dt-cell__content[data-col-index] {
			overflow: visible !important;
		}

		.party-ledger-header {
			background: linear-gradient(135deg, #f8fbff 0%, #eef5ff 100%);
			border: 1px solid #dbe7f5;
			border-radius: 18px;
			padding: 22px 24px;
			margin-bottom: 16px;
		}

		.party-ledger-header__company {
			font-size: 20px;
			font-weight: 700;
			color: #102a43;
			margin-bottom: 4px;
		}

		.party-ledger-header__title {
			font-size: 15px;
			font-weight: 600;
			color: #1f3b5b;
			margin-bottom: 10px;
		}

		.party-ledger-header__meta {
			display: flex;
			flex-wrap: wrap;
			gap: 18px;
			font-size: 12px;
			color: #52667a;
			line-height: 1.6;
		}
	`;
	document.head.appendChild(style);
}

File: ghr_custom/ghr_custom/report/party_ledger_statement/party_ledger_statement.py
import re
import frappe
from frappe import _
from frappe.utils import flt, cstr, escape_html, formatdate


def execute(filters=None):
	filters = frappe._dict(filters or {})
	validate_filters(filters)

	columns = get_columns(filters)
	message = get_report_header_html(filters)

	opening = get_opening_balance(filters)
	entries = get_gl_entries(filters)

	data = build_report_rows(filters, opening, entries)

	return columns, data, message


def build_report_rows(filters, opening, entries):
	data = []
	running_balance = opening
	show_remarks = cint_like(filters.get("show_narration"))

	if cint_like(filters.get("include_opening")):
		data.append(
			{
				"posting_date": "",
				"particulars": build_special_row_html("Opening Balance", "opening"),
				"particulars_plain": "Opening Balance",
				"voucher_subtype": "",
				"voucher_type_link": "",
				"voucher_no": "",
				"remarks": "",
				"debit": opening if opening > 0 else 0,
				"credit": abs(opening) if opening < 0 else 0,
				"balance": format_balance_number(running_balance),
				"row_kind": "opening",
				"narration_only": "",
			}
		)

	total_debit = 0.0
	total_credit = 0.0

	for gle in entries:
		debit = flt(gle.get("debit"))
		credit = flt(gle.get("credit"))

		total_debit += debit
		total_credit += credit
		running_balance += debit - credit

		narration = get_clean_narration(gle.get("remarks")) if show_remarks else ""

		data.append(
			{
				"posting_date": gle.get("posting_date"),
				"particulars": build_main_particular_html(filters, gle),
				"particulars_plain": get_main_particular_line(filters, gle),
				"voucher_subtype": get_voucher_subtype(gle),
				"voucher_type_link": gle.get("voucher_type"),
				"voucher_no": gle.get("voucher_no"),
				"remarks": narration,
				"debit": debit,
				"credit": credit,
				"balance": format_balance_number(running_balance),
				"row_kind": "normal",
				"narration_only": narration,
			}
		)

	data.append(
		{
			"posting_date": "",
			"particulars": build_special_row_html("Total", "total"),
			"particulars_plain": "Total",
			"voucher_subtype": "",
			"voucher_type_link": "",
			"voucher_no": "",
			"remarks": "",
			"debit": total_debit,
			"credit": total_credit,
			"balance": "",
			"row_kind": "total",
			"narration_only": "",
		}
	)

	data.append(
		{
			"posting_date": "",
			"particulars": build_special_row_html("Closing Balance", "closing"),
			"particulars_plain": "Closing Balance",
			"voucher_subtype": "",
			"voucher_type_link": "",
			"voucher_no": "",
			"remarks": "",
			"debit": running_balance if running_balance > 0 else 0,
			"credit": abs(running_balance) if running_balance < 0 else 0,
			"balance": format_balance_number(running_balance),
			"row_kind": "closing",
			"narration_only": "",
		}
	)

	return data


def validate_filters(filters):
	if not filters.get("company"):
		frappe.throw(_("Company is required"))

	if not filters.get("from_date"):
		frappe.throw(_("From Date is required"))

	if not filters.get("to_date"):
		frappe.throw(_("To Date is required"))

	if filters.get("from_date") > filters.get("to_date"):
		frappe.throw(_("From Date cannot be greater than To Date"))

	has_party = bool(filters.get("party"))
	has_account = bool(filters.get("account"))

	if not has_party and not has_account:
		frappe.throw(_("Please select either one Party or one Account"))

	if has_party and has_account:
		frappe.throw(_("Please select only one filter: either Party or Account, not both"))

	if has_party and not filters.get("party_type"):
		frappe.throw(_("Party Type is required when Party is selected"))


def get_columns(filters):
	show_remarks = cint_like(filters.get("show_narration"))

	columns = [
		{
			"label": _("Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 95,
		},
		{
			"label": _("Particulars"),
			"fieldname": "particulars",
			"fieldtype": "HTML",
			"width": 310,
		},
		{
			"label": _("Voucher Subtype"),
			"fieldname": "voucher_subtype",
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"label": _("Voucher No"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "voucher_type_link",
			"width": 185,
		},
		{
			"label": _("Debit"),
			"fieldname": "debit",
			"fieldtype": "Currency",
			"width": 130,
		},
		{
			"label": _("Credit"),
			"fieldname": "credit",
			"fieldtype": "Currency",
			"width": 130,
		},
		{
			"label": _("Balance"),
			"fieldname": "balance",
			"fieldtype": "Data",
			"width": 145,
		},
	]

	if show_remarks:
		columns.append(
			{
				"label": _("Remarks"),
				"fieldname": "remarks",
				"fieldtype": "HTML",
				"width": 320,
			}
		)

	return columns


def get_report_header_html(filters):
	ledger_name = filters.get("party") or filters.get("account") or ""
	ledger_type = filters.get("party_type") if filters.get("party") else "Account Ledger"

	return f"""
		<div class="party-ledger-header">
			<div class="party-ledger-header__company">{escape_html(cstr(filters.get("company")))}</div>
			<div class="party-ledger-header__title">Party Ledger Statement</div>
			<div class="party-ledger-header__meta">
				<div><strong>Ledger:</strong> {escape_html(cstr(ledger_name))}</div>
				<div><strong>Type:</strong> {escape_html(cstr(ledger_type))}</div>
				<div><strong>Period:</strong> {escape_html(formatdate(filters.get("from_date")))} to {escape_html(formatdate(filters.get("to_date")))}</div>
			</div>
		</div>
	"""


def get_opening_balance(filters):
	conditions = get_conditions(filters, before_from_date=True)

	result = frappe.db.sql(
		f"""
		select
			sum(debit) as debit,
			sum(credit) as credit
		from `tabGL Entry`
		where docstatus < 2
			and ifnull(is_cancelled, 0) = 0
			{conditions}
		""",
		filters,
		as_dict=True,
	)

	debit = flt(result[0].get("debit")) if result else 0
	credit = flt(result[0].get("credit")) if result else 0
	return debit - credit


def get_gl_entries(filters):
	conditions = get_conditions(filters, before_from_date=False)

	return frappe.db.sql(
		f"""
		select
			name,
			posting_date,
			voucher_type,
			voucher_no,
			account,
			party_type,
			party,
			against,
			remarks,
			debit,
			credit,
			cost_center,
			project,
			is_cancelled
		from `tabGL Entry`
		where docstatus < 2
			and ifnull(is_cancelled, 0) = 0
			{conditions}
		order by posting_date, creation, name
		""",
		filters,
		as_dict=True,
	)


def get_conditions(filters, before_from_date=False):
	conditions = ["and company = %(company)s"]

	if before_from_date:
		conditions.append("and posting_date < %(from_date)s")
	else:
		conditions.append("and posting_date between %(from_date)s and %(to_date)s")

	if filters.get("party"):
		conditions.append("and party_type = %(party_type)s")
		conditions.append("and party = %(party)s")

	if filters.get("account"):
		conditions.append("and account = %(account)s")

	if filters.get("cost_center"):
		conditions.append("and cost_center = %(cost_center)s")

	if filters.get("project"):
		conditions.append("and project = %(project)s")

	return "\n".join(conditions)


def build_main_particular_html(filters, gle):
	main_line = escape_html(get_main_particular_line(filters, gle))
	return f"<div class='pls-main'>{main_line}</div>"


def build_special_row_html(label, row_type):
	css_map = {
		"opening": "pls-special pls-opening",
		"total": "pls-special pls-total",
		"closing": "pls-special pls-closing",
	}
	css_class = css_map.get(row_type, "pls-special")
	return f"<div class='{css_class}'>{escape_html(label)}</div>"


def get_main_particular_line(filters, gle):
	if filters.get("party"):
		counter_name = get_against_label(gle)
		prefix = "To" if flt(gle.get("debit")) > 0 else "By"
		return f"{prefix} {counter_name}"

	if filters.get("account"):
		counter_name = get_party_or_against_label(gle)
		prefix = "To" if flt(gle.get("debit")) > 0 else "By"
		return f"{prefix} {counter_name}"

	return get_against_label(gle)


def get_against_label(gle):
	against = cstr(gle.get("against")).strip()
	if against:
		return against

	account = cstr(gle.get("account")).strip()
	if account:
		return account

	return "-"


def get_party_or_against_label(gle):
	party = cstr(gle.get("party")).strip()
	if party:
		return party

	against = cstr(gle.get("against")).strip()
	if against:
		return against

	account = cstr(gle.get("account")).strip()
	if account:
		return account

	return "-"


def get_voucher_subtype(gle):
	voucher_type = cstr(gle.get("voucher_type"))
	voucher_no = cstr(gle.get("voucher_no"))

	if not voucher_type or not voucher_no:
		return voucher_type

	try:
		if voucher_type == "Payment Entry":
			payment_type = frappe.get_cached_value("Payment Entry", voucher_no, "payment_type")
			return payment_type or voucher_type

		if voucher_type == "Journal Entry":
			je_type = frappe.get_cached_value("Journal Entry", voucher_no, "voucher_type")
			return je_type or voucher_type

		if voucher_type == "Purchase Invoice":
			is_return = frappe.get_cached_value("Purchase Invoice", voucher_no, "is_return")
			return "Debit Note" if is_return else "Purchase Invoice"

		if voucher_type == "Sales Invoice":
			is_return = frappe.get_cached_value("Sales Invoice", voucher_no, "is_return")
			return "Credit Note" if is_return else "Sales Invoice"

	except Exception:
		pass

	return voucher_type


def get_clean_narration(raw_remarks):
	narration = cstr(raw_remarks).strip()
	if not narration:
		return ""

	normalized = re.sub(r"[\s\.\-_]+", " ", narration).strip().lower()

	if normalized in {"no remarks", "no remark", "remarks", "remark", "n a", "na", "nil", "-"}:
		return ""

	return narration


def format_balance_number(amount):
	amount = flt(amount)
	if not amount:
		return "0.00"
	return f"{amount:,.2f}"


def cint_like(value):
	return 1 if str(value) in ("1", "true", "True") else 0


def format_print_date_ddyymmdd(date_value):
	if not date_value:
		return ""
	return formatdate(date_value, "dd-mm-yy")


@frappe.whitelist()
def get_print_data(filters=None):
	if isinstance(filters, str):
		filters = frappe.parse_json(filters)

	filters = frappe._dict(filters or {})
	validate_filters(filters)

	opening = get_opening_balance(filters)
	entries = get_gl_entries(filters)
	data = build_report_rows(filters, opening, entries)

	print_rows = []
	for row in data:
		row = frappe._dict(row)

		print_rows.append(
			{
				**row,
				"posting_date": format_print_date_ddyymmdd(row.get("posting_date")),
			}
		)

		narration = row.get("narration_only") or row.get("remarks") or ""
		if row.get("row_kind") == "normal" and narration:
			print_rows.append(
				{
					"posting_date": "",
					"particulars": "",
					"particulars_plain": "",
					"voucher_subtype": "",
					"voucher_type_link": "",
					"voucher_no": "",
					"remarks": "",
					"debit": "",
					"credit": "",
					"balance": "",
					"row_kind": "narration",
					"narration_only": narration,
				}
			)

	return {
		"company": filters.get("company"),
		"from_date": format_print_date_ddyymmdd(filters.get("from_date")),
		"to_date": format_print_date_ddyymmdd(filters.get("to_date")),
		"party_type": filters.get("party_type") or "",
		"ledger_name": filters.get("party") or filters.get("account") or "",
		"rows": print_rows,
	}


@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def get_party_query(doctype, txt, searchfield, start, page_len, filters):
	filters = frappe._dict(filters or {})

	party_type = filters.get("party_type")
	company = filters.get("company")

	if not party_type or not company:
		return []

	return frappe.db.sql(
		"""
		select distinct gle.party
		from `tabGL Entry` gle
		where gle.party_type = %(party_type)s
			and gle.company = %(company)s
			and ifnull(gle.party, '') != ''
			and ifnull(gle.is_cancelled, 0) = 0
			and gle.docstatus < 2
			and gle.party like %(txt)s
		order by gle.party
		limit %(start)s, %(page_len)s
		""",
		{
			"party_type": party_type,
			"company": company,
			"txt": f"%{txt}%",
			"start": start,
			"page_len": page_len,
		},
		as_list=True,
	)
